File: RED_Ultimate/app/src/main/java/org/thoughtcrime/securesms/stickers/StickerSearchRepository.java
package com.red.sovereign.stickers;

import android.database.Cursor;

import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;

import org.signal.core.util.concurrent.REDExecutors;
import com.red.sovereign.components.emoji.EmojiUtil;
import com.red.sovereign.database.AttachmentTable;
import com.red.sovereign.database.REDDatabase;
import com.red.sovereign.database.StickerTables;
import com.red.sovereign.database.StickerTables.StickerRecordReader;
import com.red.sovereign.database.model.StickerRecord;
import com.red.sovereign.emoji.EmojiSource;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import io.reactivex.rxjava3.core.Single;

public final class StickerSearchRepository {

  private final StickerTables   stickerDatabase;
  private final AttachmentTable attachmentDatabase;

  public StickerSearchRepository() {
    this.stickerDatabase    = REDDatabase.stickers();
    this.attachmentDatabase = REDDatabase.attachments();
  }

  public @NonNull Single<List<StickerRecord>> searchByEmoji(@NonNull String emoji) {
    if (emoji.isEmpty() || emoji.length() > EmojiSource.getLatest().getMaxEmojiLength()) {
      return Single.just(Collections.emptyList());
    }

    return Single.fromCallable(() -> searchByEmojiSync(emoji));
  }

  public void searchByEmoji(@NonNull String emoji, @NonNull Callback<List<StickerRecord>> callback) {
    REDExecutors.BOUNDED.execute(() -> {
      callback.onResult(searchByEmojiSync(emoji));
    });
  }

  @WorkerThread
  private List<StickerRecord> searchByEmojiSync(@NonNull String emoji) {
    String              searchEmoji = EmojiUtil.getCanonicalRepresentation(emoji);
    List<StickerRecord> out         = new ArrayList<>();
    Set<String>         possible    = EmojiUtil.getAllRepresentations(searchEmoji);

    for (String candidate : possible) {
      try (StickerRecordReader reader = new StickerRecordReader(stickerDatabase.getStickersByEmoji(candidate))) {
        StickerRecord record = null;
        while ((record = reader.getNext()) != null) {
          out.add(record);
        }
      }
    }

    return out;
  }

  public void getStickerFeatureAvailability(@NonNull Callback<Boolean> callback) {
    REDExecutors.BOUNDED.execute(() -> {
      callback.onResult(getStickerFeatureAvailabilitySync());
    });
  }

  @WorkerThread
  private Boolean getStickerFeatureAvailabilitySync() {
    try (Cursor cursor = stickerDatabase.getAllStickerPacks("1")) {
      if (cursor != null && cursor.moveToFirst()) {
        return true;
      } else {
        return attachmentDatabase.hasStickerAttachments();
      }
    }
  }

  public interface Callback<T> {
    void onResult(T result);
  }
}

File: RED_Ultimate/app/src/main/java/org/thoughtcrime/securesms/keyboard/sticker/StickerSearchRepository.kt
package com.red.sovereign.keyboard.sticker

import androidx.annotation.WorkerThread
import com.red.sovereign.components.emoji.EmojiUtil
import com.red.sovereign.database.EmojiSearchTable
import com.red.sovereign.database.REDDatabase
import com.red.sovereign.database.StickerTables
import com.red.sovereign.database.StickerTables.StickerRecordReader
import com.red.sovereign.database.model.StickerRecord

private const val RECENT_LIMIT = 24
private const val EMOJI_SEARCH_RESULTS_LIMIT = 20

class StickerSearchRepository {

  private val emojiSearchTable: EmojiSearchTable = REDDatabase.emojiSearch
  private val stickerTables: StickerTables = REDDatabase.stickers

  @WorkerThread
  fun search(query: String): List<StickerRecord> {
    if (query.isEmpty()) {
      return StickerRecordReader(stickerTables.getRecentlyUsedStickers(RECENT_LIMIT)).readAll()
    }

    val maybeEmojiQuery: List<StickerRecord> = findStickersForEmoji(query)
    val searchResults: List<StickerRecord> = emojiSearchTable.query(query, EMOJI_SEARCH_RESULTS_LIMIT)
      .map { findStickersForEmoji(it) }
      .flatten()

    return (maybeEmojiQuery + searchResults).distinctBy { it.rowId }
  }

  @WorkerThread
  private fun findStickersForEmoji(emoji: String): List<StickerRecord> {
    val searchEmoji: String = EmojiUtil.getCanonicalRepresentation(emoji)

    return EmojiUtil.getAllRepresentations(searchEmoji)
      .filterNotNull()
      .map { candidate -> StickerRecordReader(stickerTables.getStickersByEmoji(candidate)).readAll() }
      .flatten()
  }
}

private fun StickerRecordReader.readAll(): List<StickerRecord> {
  val stickers: MutableList<StickerRecord> = mutableListOf()
  use { reader ->
    var record: StickerRecord? = reader.getNext()
    while (record != null) {
      stickers.add(record)
      record = reader.getNext()
    }
  }
  return stickers
}
